src/polyglot-notebooks-vscode-common/src/vscodeNotebookManagement.ts

```diff
@@ -31,13 +31,44 @@ export async function replaceNotebookCellMetadata(notebookUri: vscode.Uri, cellI
 }
 
 export async function replaceNotebookMetadata(notebookUri: vscode.Uri, documentMetadata: { [key: string]: any }): Promise<boolean> {
-    const notebookEdit = vscode.NotebookEdit.updateNotebookMetadata(documentMetadata);
-    const edit = new vscode.WorkspaceEdit();
-    edit.set(notebookUri, [notebookEdit]);
-    const succeeded = await vscode.workspace.applyEdit(edit);
-    return succeeded;
+    const notebook = vscode.workspace.notebookDocuments.find(d => d.uri === notebookUri);
+    if (notebook) {
+        const metadata = notebook.metadata;
+        const shouldUpdate = compare(metadata, documentMetadata);
+        if (shouldUpdate) {
+            const notebookEdit = vscode.NotebookEdit.updateNotebookMetadata(documentMetadata);
+            const edit = new vscode.WorkspaceEdit();
+            edit.set(notebookUri, [notebookEdit]);
+            const succeeded = await vscode.workspace.applyEdit(edit);
+            return succeeded;
+        }
+        return false;
+    }
+    return false;
 }
 
+function compare(object1: { [key: string]: any }, object2: { [key: string]: any }) {
+    const isObject = (object: any) => {
+        return object !== null && typeof object === 'object';
+    }
+    const objectKeys1 = Object.keys(object1);
+    const objectKeys2 = Object.keys(object2);
+    if (objectKeys1.length !== objectKeys2.length) {
+        return false;
+    }
+    for (const key of objectKeys1) {
+        const value1 = object1[key];
+        const value2 = object2[key];
+        const bothAreObjects = isObject(value1) && isObject(value2);
+        if ((bothAreObjects && !compare(value1, value2)) || (!bothAreObjects && value1 === value2)) {
```

Comment:
I don't have much experience with TypeScript and how the JSON comparison works, but shouldn't the RHS for the `||` operator be `!bothAreObjects && value1 != value2` then `return false`? In other words, if both are non-objects but yet equal, shouldn't this `return true`?

Also, should this `return false` if one of the values is an object and the other isn't?

Reply:
changed